cmd/server-mux.go

```diff
@@ -113,30 +113,46 @@ func (c *ConnMux) PeekProtocol() (string, error) {
 	return protocolTLS, nil
 }
 
-// Read - streams the ConnMux buffer when reset flag is activated, otherwise
-// streams from the incoming network connection
-func (c *ConnMux) Read(b []byte) (n int, e error) {
-	// Push read deadline
-	c.Conn.SetReadDeadline(time.Now().Add(defaultTCPReadTimeout))
-
-	// Update server's connection statistics
-	defer func() {
-		globalConnStats.incInputBytes(n)
-	}()
+// Read reads from the tcp session for data sent by
+// the client, additionally sets deadline for 15 secs
+// after each successful read. Deadline cancels and
+// returns error if the client does not send any
+// data in 15 secs. Also keeps track of the total
+// bytes received from the client.
+func (c *ConnMux) Read(b []byte) (n int, err error) {
+	n, err = c.peeker.Read(b)
+	if err != nil {
+		return n, err
+	}
+
+	// Read deadline was already set previously, set again
+	// after a successful read operation for future read
+	// operations.
+	c.Conn.SetReadDeadline(UTCNow().Add(defaultTCPReadTimeout))
+
+	// Update total incoming number of bytes.
+	globalConnStats.incInputBytes(n)
 
-	return c.peeker.Read(b)
+	// Success.
+	return n, nil
 }
 
-func (c *ConnMux) Write(b []byte) (n int, e error) {
-	// Update server's connection statistics
-	defer func() {
-		globalConnStats.incOutputBytes(n)
-	}()
-	// Run the underlying net.Conn Write() func
-	return c.Conn.Write(b)
+// Write to the client over a tcp session, additionally
+// keeps track of the total bytes written by the server.
+func (c *ConnMux) Write(b []byte) (n int, err error) {
+	n, err = c.Conn.Write(b)
+	if err != nil {
+		return n, err
+	}
+
+	// Update total outgoing number of bytes.
+	globalConnStats.incOutputBytes(n)
```

Comment:
> Yes Only success is counted because we don't need remember data which we didn't write to disk.

I see. But our stats (input/output bytes) won't look reliable that way. Besides, if a user wants to see data written in the disk, he won't need the exact values, the approximation will be sufficient for him.